src/cargo/sources/registry/http_remote.rs

```diff
@@ -0,0 +1,666 @@
+    }
+
+    fn download(&mut self, pkg: PackageId, checksum: &str) -> CargoResult<MaybeLock> {
+        let registry_config = loop {
+            match self.config()? {
+                Poll::Pending => self.block_until_ready()?,
+                Poll::Ready(cfg) => break cfg.unwrap(),
+            }
+        };
+        download::download(
+            &self.cache_path,
+            &self.config,
+            pkg,
+            checksum,
+            registry_config,
+        )
+    }
+
+    fn finish_download(
+        &mut self,
+        pkg: PackageId,
+        checksum: &str,
+        data: &[u8],
+    ) -> CargoResult<File> {
+        download::finish_download(&self.cache_path, &self.config, pkg, checksum, data)
+    }
+
+    fn is_crate_downloaded(&self, pkg: PackageId) -> bool {
+        download::is_crate_downloaded(&self.cache_path, &self.config, pkg)
+    }
+
+    fn block_until_ready(&mut self) -> CargoResult<()> {
+        let initial_pending_count = self.downloads.pending.len();
+        trace!(
+            "block_until_ready: {} transfers pending",
+            initial_pending_count
+        );
+
+        loop {
+            self.handle_completed_downloads()?;
+
+            let remaining_in_multi = tls::set(&self.downloads, || {
+                self.multi
+                    .perform()
+                    .with_context(|| "failed to perform http requests")
+            })?;
+            trace!("{} transfers remaining", remaining_in_multi);
+
+            if remaining_in_multi == 0 {
+                return Ok(());
+            }
+
+            // We have no more replies to provide the caller with,
+            // so we need to wait until cURL has something new for us.
+            let timeout = self
+                .multi
+                .get_timeout()?
+                .unwrap_or_else(|| Duration::new(5, 0));
+            self.multi
+                .wait(&mut [], timeout)
+                .with_context(|| "failed to wait on curl `Multi`")?;
+        }
+    }
+}
+
+impl<'cfg> Downloads<'cfg> {
+    fn progress(&self, token: usize, total: u64, cur: u64) -> bool {
+        let dl = &self.pending[&token].0;
+        dl.total.set(total);
+        dl.current.set(cur);
+        true
+    }
+
+    fn tick(&self) -> CargoResult<()> {
+        let mut progress = self.progress.borrow_mut();
+        let progress = progress.as_mut().unwrap();
+
+        progress.tick(
+            self.downloads_finished,
+            self.downloads_finished + self.pending.len(),
+            "",
+        )
+    }
+}
+
+mod tls {
+    use super::Downloads;
+    use std::cell::Cell;
+
+    thread_local!(static PTR: Cell<usize> = Cell::new(0));
+
+    pub(crate) fn with<R>(f: impl FnOnce(Option<&Downloads<'_>>) -> R) -> R {
+        let ptr = PTR.with(|p| p.get());
+        if ptr == 0 {
+            f(None)
+        } else {
+            // Safety: * `ptr` is only set by `set` below which ensures the type is correct.
```

Comment:
It's not actually the type I was worried about here. It's the lifetime. What prevents us using downloads after it's been freed? It took me trying to write a counter example to realize why this is safe. `ptr` is only set during the call back in `set`. The way we use it is a little confusing, because curl needs us to have "callbacks at a distance". But our callbacks are only called within the call back to set.

I don't know how to document this better, nor do I think it has to be done before merge.